### PR TITLE
fix a bug in ClassPositionsOfLowerCentralSeries

### PR DESCRIPTION
**Pull request by ThomasBreuer**
This pull request is intended to deal with issue #3321.

The proposed changes are as follows.

1. fixed errors in
   `ClassPositionsOfLowerCentralSeries`, `ClassPositionsOfSolvableResiduum`, `ClassPositionsOfSupersolvableResiduum`, which may lead to wrong results

2. avoid calling `ClassPositionsOfNormalSubgroups` in situations where this may lead to a huge list
   (the worst cases are large elementary abelian groups),
   in `IsSimpleCharacterTable`, `ClassPositionsOfMinimalNormalSubgroups`, `ClassPositionsOfFittingSubgroup`, `ClassPositionsOfSolvableRadical`, `ClassPositionsOfSolvableResiduum`, `ClassPositionsOfSupersolvableResiduum`, `IsPSolvableCharacterTableOp`

3. added extensive tests for the `ClassPositionsOf...` functions, which would not be reasonable without the changes under 2

4. fixed an `ElementaryAbelianSeries` method such that `fail` is returned in the case of nonsolvable groups, instead of calling `Error`; note that the documentation promises `fail` in this case

5. as a consequence of 4., `GroupSeriesByNiceMonomorphism` is now aware of `fail` results

### CODE REVIEW

File: lib/ctbl.gi [hunk 1]
```diff
@@ -1943,12 +1952,10 @@ InstallMethod( ClassPositionsOfMinimalNormalSubgroups,
           minimal,   # list of minimal kernels
           k;         # one kernel
 
-    # Every normal subgroup is an intersection of kernels of characters,
-    # so maximal normal subgroups are kernels of irreducible characters.
-    normal:= Set( ClassPositionsOfNormalSubgroups( tbl ) );
+    normal:= Set( List( [ 2 .. NrConjugacyClasses( tbl ) ],
```
Comment (fingolfin) [inline lib/ctbl.gi]: Minor remark: `Set( List( l, f ) )` is equivalent to (but slower than) `Set( l, f )`

File: lib/ctbl.gi [hunk 2]
```diff
@@ -2185,70 +2204,64 @@ InstallMethod( ClassPositionsOfSolvableRadical,
...
-    # Compute the structure constants $a_{x,\overline{x},g}$ with $g$ and $x$
-    # in $K_2$.
+    # Compute the structure constants $a_{g,x,x}$ with $g$ and $x$ in $K_2$.
     # Put them into a matrix, the rows indexed by $g$, the columns by $x$.
-    inv:= InverseClasses( tbl );
     mat:= List( K, x -> [] );
     for i in [ 2 .. Length( K ) ] do
       for j in K do
-        mat[i][j]:= ClassMultiplicationCoefficient( tbl, K[i], j, inv[j] );
+        mat[i,j]:= ClassMultiplicationCoefficient( tbl, K[i], j, j );
```
Comment (fingolfin) [inline lib/ctbl.gi]: I have no idea whether this is correct or not, so I'll just trust that you know what you are doing :-).

File: lib/ctbl.gi [hunk 3]
```diff
@@ -2185,70 +2204,64 @@ InstallMethod( ClassPositionsOfSolvableRadical,
...
       if Length( new ) = Length( K ) then
-        running:= false;
+        break;
       else
-        mat:= mat{ new };
-        K:= K{ new };
-        Add( series, new );
+        mat:= mat{ List( new, i -> Position( K, i ) ) };
+        K:= new;
+        Add( series, K );
```
Comment (fingolfin) [inline lib/ctbl.gi]: Again, I'll trust that you know what you are doing.

File: lib/ctbl.gi (hunk 1 above)
Comment (ThomasBreuer) [reply lib/ctbl.gi]: @fingolfin Interesting, the two argument version of `Set` (resp. `SSortedList`) seems to be undocumented. Should we document it?

Comment (fingolfin) [reply lib/ctbl.gi]: Ah, I was not aware that it is undocumented. Yes, I think it should be documented!

**Comment by fingolfin** [comment]
Backported to stable-4.10 in commits 565e12fd4fb48091e65f16c9a3652462bbe4121d, 0ebdb142e58593aef1eddc71dc8389790f7e2b63, 55fba348c8185aa6ee5e58f5517cdfbb06ac8b59

**Pull request merged**